### ISSUE TITLE
Outdated root certificates for cURL on old macOS

### ISSUE THREAD

**Issue by dotnetCarpenter** [fink/scripts#60]
Hi all.

I am installing Fink for the first time on macOS 10.13 High Sierra. I think most people on old macOS'es will run into the same issue as I did. Which is that the root certificates are too old and cURL will refuse to download any needed packages.

For the initial step, I used [Install Fink.tool][1] and added `-k` the `curl` line, which side steps cURL's verification. However, cURL [maintains updated root certificates, extracted from Firefox][2] and it would probably be better to patch [Intall Fink.tool line 78][3] to allow the user to set `--cacert` for cURL. _On my system it is not possible to programmatically verify the cert file with sha256 since macOS 10.13 does not have `sha256sum`._

Let me know if a patch is accepted for this and what the User Experience (UX) should be. E.i. should it be an argument to _Install Fink.tool_ or a prompt?

[1]: https://github.com/fink/scripts/releases/tag/setup/0.45.6
[2]: https://curl.se/docs/caextract.html
[3]: https://github.com/fink/scripts/blob/setup/0.45.6/srcinstaller/Install%20Fink.tool#L78

**Comment by TheSin-** [comment]
IIRC that is why I built the ca-certificates package, it fixes these issues.

 i   ca-certificates  20210119-2   Common CA certificates

**Comment by dotnetCarpenter** [comment]
@TheSin- I don't know the _ca-certificates package_. Is it meant to be installed via `fink`?

**Comment by TheSin-** [comment]
yes it is, and fixes the old certs for wget and curl.  If it was part of the bootstrap it would be installed at the start of the bootstrap to avoid such errors and allow proper verification of certs

**Comment by dotnetCarpenter** [comment]
I might misunderstand what you are saying.
But the issue is that `FinkSourceDLP="https://downloads.sourceforge.net/project/fink/fink/${FinkVersion}/${FinkFileName}"` can not be downloaded without updated root certificates.

When you say _bootstrap_ - do you mean the _Install Fink.tool_ bash script or the _fink_0.x.x-141_darwin-x86_64.deb_ file?

**Comment by TheSin-** [comment]
the deb file is built on your system, but before it's build it needs to install a variety of other packages, like ncurses.  That is what we call bootstrapping, it builds and installs everything required to install fink, think of it like pre depends.

We could make the ca-certificates package be one of the first thinks installed to voice ssl cert validation errors during the rest of the boot strap.  I don't have such an old setup to test with.  But in theory it's possible, and this would allow us to make it more robust and future proof.

**Comment by TheSin-** [comment]
I'm sure @dak180 will chime in, he is way more familiar with the script as I normally do everything manually from git directly.

**Comment by dotnetCarpenter** [comment]
I'm afraid that won't work in my situation. You can not download anything without updated root certificates unless you disable cURL verification. It must be part of the initial manual download. Or be explained to the user when executing [Install Fink.tool](https://github.com/fink/scripts/releases/tag/setup/0.45.6) for the first time.

Currently the error message is:

```
curl: (60) SSL certificate problem: certificate has expired
More details here: https://curl.haxx.se/docs/sslcerts.html
```

**Comment by TheSin-** [comment]
oh I see it's getting the binary depends like java and XDarwin first.

Okay I'll let @dak180 handle this, he'll figure it out for you.

**Comment by dak180** [comment]
I have an idea for fixing this but I will not have a chance to do anything with it until monday.

**Comment by dotnetCarpenter** [comment]
@dak180 As far as I understand, disabling verification in cURL calls will still result in an encrypted connection. Only the certificate authenticity is not validated.

**Comment by dotnetCarpenter** [comment]
> uname -r | cut -d. -f1

17

Darwin Kernel Version 17.7.0

**Comment by dotnetCarpenter** [comment]
I just realised that a simpler method is to download the newest cacert.pem and then create or add to _.curlrc_:

```
--cacert [location/of/cacert.pem]
```

This also works for the `curl` requests that `fink` makes in other places.

**Comment by nieder** [comment]
chicken-egg problem. Can't download `cacert.pem` from [cURL web site](https://curl.se) w/out a working curl, which needs `cacert.pem`.
`Install Fink.tool` gets downloaded as a tgz file. Could put cacert.pem into the repo, add it to the tgz, and then this script would have it. Does `--cacert` take relative paths?

But the Fink bootstrap method won't be able to use this `cacert.pem` unless patched (passing CURL_CA_BUNDLE from `Install Fink.tool` will be ignored). Let's see what @dak180 's plan is...

**Comment by dotnetCarpenter** [comment]
>Does --cacert take relative paths?

Too late for me to test with the cURL version that comes with 10.13. But the newest cURL (curl 7.86.0) does support relative paths as argument to `--cacert`.

Here is a link to the man page: https://curl.se/docs/manpage.html#--cacert

Absolute path with old root certificates:
```
curl --cacert /etc/ssl/cert.pem -LI https://mandoc.bsd.lv/snapshots/mandoc-1.14.6.tar.gz
curl: (60) SSL: certificate verification failed (result: 5)
```

Relative path with new root certificates:
```
curl --cacert "../../rootcerts.pem" -LI https://mandoc.bsd.lv/snapshots/mandoc-1.14.6.tar.gz
HTTP/1.1 200 OK
```

**Comment by dotnetCarpenter** [comment]
>But the Fink bootstrap method won't be able to use this cacert.pem unless patched (passing CURL_CA_BUNDLE from Install Fink.tool will be ignored).

@nieder I would think that using an environment variable, `CURL_CA_BUNDLE`, or having `--cacert` in _.curlrc_ would work. Both would do the same for all cURL calls, regardless of from where `curl` is called.
Is that not the case?

**Comment by dak180** [comment]
Fixed in ab862209c28e012f9ecb4030a179bdec84ee3b16.
